cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManager.java

```diff
@@ -0,0 +1,422 @@
+    ApplicationId applicationId = new ApplicationId(program.getNamespace(), program.getApplication(),
+                                                    program.getVersion());
+    return new ProgramId(applicationId, ProgramType.valueOf(program.getType()), program.getName());
+  }
+
+  @Nullable
+  private String getCursorResponse(MetadataSearchResponse searchResponse) {
+    List<String> cursors = searchResponse.getCursors();
+    if (cursors == null || cursors.isEmpty()) {
+      return null;
+    }
+    return cursors.get(0);
+  }
+
+  private ApplicationId getApplicationId(MetadataEntity metadataEntity) {
+    return new ApplicationId(metadataEntity.getValue(MetadataEntity.NAMESPACE),
+                             metadataEntity.getValue(MetadataEntity.APPLICATION),
+                             metadataEntity.getValue(MetadataEntity.VERSION));
+  }
+
+  private void enableCapability(String capability) {
+    capabilityStatusMap.put(capability, CapabilityStatus.ENABLED);
+    LOG.debug("Capability {} enabled.", capability);
+  }
+
+  private void disableCapability(String capability) {
+    //mark as disabled to prevent further runs
+    capabilityStatusMap.put(capability, CapabilityStatus.DISABLED);
+    //stop pipelines
+    try {
+      stopPipelines(capability);
+    } catch (Exception ex) {
+      LOG.error("Stopping pipelines failed for capability {} with exception {}", capability, ex);
+    }
+    //programs(services) will be stopped by SystemProgramManagementService
+    LOG.debug("Capability {} disabled.", capability);
+  }
+
+  private void deleteCapability(CapabilityConfig capabilityConfig) {
+    String capability = capabilityConfig.getCapability();
+    if (capabilityStatusMap.get(capability) == CapabilityStatus.ENABLED) {
+      LOG.error("Deleting capability {} failed. Capability should be disabled before deleting.", capability);
+      return;
+    }
+    //delete pipelines
+    try {
+      deletePipelines(capability);
+    } catch (Exception ex) {
+      LOG.error("Deleting pipelines failed for capability {} with exception {}", capability, ex);
+    }
+    //delete applications
+    for (SystemApplication application : capabilityConfig.getApplications()) {
+      ApplicationId applicationId = new ApplicationId(application.getNamespace(), application.getName(),
+                                                      application.getVersion());
+      try {
+        applicationLifecycleService.removeApplication(applicationId);
+      } catch (Exception exception) {
+        LOG.error("Deleting application {} failed with exception {}", applicationId, exception);
+      }
+    }
+    capabilityStatusMap.remove(capability);
+    LOG.debug("Capability {} deleted.", capability);
+  }
+
+  private void deletePipelines(String capability) throws Exception {
+    doForAllApps(capability, this::deleteAppWithRetry);
+  }
+
+  private void stopPipelines(String capability) throws Exception {
```

Comment:
Shouldn't call pipeline. CDAP platform itself has no concept of pipeline. Call it stopAllPrograms

Reply:
Renamed.